### ISSUE TITLE
create a zoomable map with tiles using leafletsjs

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#26]
We should explore a new mechanism of loading our Mars maps and implement it in the next two versions of mars-sim.

I've just found leaflet, an open-source JavaScript library for creating interactive maps that JavaFX's Webview/WebEngine can use. 

See http://leafletjs.com/

Since leaflet is built to be mobile friendly, it makes it easier for us to create mars-sim for android or mars-sim for web that will leverage its versatility.

Also, I found someone used GDAL2Tiles with python to split an excessively large giga-pixel image into tiles.  It's a astronomical picture of the Central parts of the Milky Way with a catalog of 84 million stars.
http://build-failed.blogspot.com/2012/11/zoomable-image-with-leaflet.html

Next he uses MapTiler to process those tiles from GDAL2Tiles.
http://www.maptiler.com/

At last he uses leaflet for loading the zoomable image in a website.

Btw, we can use gpx for recording tracks on Mars : 
http://www.topografix.com/gpx_resources.asp
http://wiki.openstreetmap.org/wiki/GPX

I'll upload GpxParseDemo.java into mars-sim-ui's /javafx/demo. It was written by Geritt for integrating the use of leafletjs.js in Java via JavaFX's WebEngine.
https://harmoniccode.blogspot.com/
https://github.com/HanSolo/tilesfx

I believe we're not too far from writing/refactoring a few new java class for loading new Mars maps using leafletjs.

Let me know if you want to tackle on this.

**Comment by mokun** [comment]
See also this stackexchange discussion at
http://gis.stackexchange.com/questions/199987/non-earth-maps-tile-server-and-scale-in-leafletjs#200026

Interestingly, someone wrote a lealfet plugin for displaying celestial bodies from Kerbal Space Program :
https://github.com/saik0/Leaflet.KSP

**Comment by larsnaesbye** [comment]
Question is if we can piggyback on top of classic map service definitions, like WMTS. I'm lucky to work with dispatching WMTS requests to backend renderers, so I might look into this.

**Comment by mokun** [comment]
I don't know how  WMTS requests work. It would require online access.

We can prepare level 1-3 map tiles by using leaftlet, GDAL2Tiles and MapTiller and distribute these tiles by default in the jar package. 

We can code it in such a way that If users want to zoom in further, mars-sim will do WMTS request to load up level 4-10 map tiles, etc.

**Comment by larsnaesbye** [comment]
You're right. Full-blown WMTS might be overkill. leaflet, GDAL2Tiles and MapTiller it is.

**Comment by mokun** [comment]
Mapbox has a Leaflet plugin at https://github.com/mapbox/mapbox.js/

**Comment by mokun** [comment]
@Urwumpe @bevans2000 @larsnaesbye 

I'm looking at the examples at https://leafletjs.com/examples.html. 

In the past with JavaFX, we were briefly capable of running some html 5.0/5.1 codes and it may even be capable of embedding JS codes in mars-sim and using [GeoJSON](https://leafletjs.com/examples/geojson/) with Leaflet. It would have been powerful but may still be cumbersome since it's not Java-native and will have to switch between Java and JS a lot.

**Comment by larsnaesbye** [comment]
Is there a way to make Leaflet aware of Mars projections? I assume we'll get aberrations if we use a terrestrial geoid?

**Comment by mokun** [comment]
> Is there a way to make Leaflet aware of Mars projections? 

See this at 
https://www.openplanetary.org/opm-basemaps/opm-mars-basemap-v0-2.
